lib/segment/src/segment_constructor/segment_constructor_base.rs

```diff
@@ -1315,7 +1321,134 @@ pub fn migrate_rocksdb_sparse_vector_storage_to_mmap(
+    // Update storage type
+    // TODO(in-memory-mmap-payload-storage): once in-memory mmap payload storage is merged, select correct type here
+    segment_state.config.payload_storage_type = if old_storage.is_on_disk() {
+        PayloadStorageType::Mmap
+    } else {
+        PayloadStorageType::InRamMmap
+    };
+    Segment::save_state(segment_state, path)?;
+
+    // Destroy persisted RocksDB payload data
+    match old_storage {
+        PayloadStorageEnum::SimplePayloadStorage(storage) => storage.destroy()?,
+        PayloadStorageEnum::OnDiskPayloadStorage(storage) => storage.destroy()?,
+        PayloadStorageEnum::MmapPayloadStorage(_) => {
+            unreachable!("unexpected payload storage type")
+        }
+        #[cfg(feature = "testing")]
+        PayloadStorageEnum::InMemoryPayloadStorage(_) => {
+            unreachable!("unexpected payload storage type")
+        }
+    }
+
+    // Also update config in already loaded segment
+    segment.segment_config = segment_state.config.clone();
+
+    Ok(())
+}
+
+/// Migrate a RocksDB based payload storage storage into the mmap format
+///
+/// Creates a new payload storage on top of memory maps, and copies all payloads
+/// from the RocksDB based storage into it. The new payload storage is returned.
+///
+/// Warning: the old payload storage is not destroyed, so it must be done manually
+#[cfg(feature = "rocksdb")]
+pub fn migrate_rocksdb_payload_storage_to_mmap(
+    old_storage: &PayloadStorageEnum,
+    segment_path: &Path,
+) -> OperationResult<PayloadStorageEnum> {
+    use common::counter::hardware_counter::HardwareCounterCell;
+
+    use crate::payload_storage::PayloadStorage;
+    use crate::payload_storage::mmap_payload_storage::find_storage_files;
+
+    log::info!(
+        "Migrating {} bytes of payload storage from RocksDB into new format",
+        old_storage.get_storage_size_bytes().unwrap_or(0),
+    );
+
+    fn migrate(
+        old_storage: &PayloadStorageEnum,
+        segment_path: &Path,
+    ) -> OperationResult<PayloadStorageEnum> {
+        // Construct mmap based payload storage
+        let mut new_storage = PayloadStorageEnum::from(MmapPayloadStorage::open_or_create(
+            segment_path.to_path_buf(),
+            !old_storage.is_on_disk(),
+        )?);
+
+        // Copy all payloads and deletes into new storage
+        let hw_counter = HardwareCounterCell::disposable();
+        old_storage.iter(
+            |internal_id, payload| {
+                new_storage.set(internal_id, payload, &hw_counter)?;
+                Ok(true)
+            },
+            &hw_counter,
+        )?;
+
+        // Flush new storage
+        new_storage.flusher()()?;
+
+        Ok(new_storage)
+    }
+
+    let new_storage = match migrate(old_storage, segment_path) {
+        Ok(new_storage) => new_storage,
+        // On migration error, clean up and remove all new storage files
+        Err(err) => {
+            let files = find_storage_files(segment_path);
```

Comment:
this deletes files, but do we care about directories?

Reply:
Also, what happens if migration is aborted? e.g. User tries to cancel the migration by killing the process.

I think the files will still be there, so next time it `open_or_create` it will open instead, but everything will be rewritten... It should be fine.

I guess I answered my own question 😄

Reply:
I now drop the full directory: https://github.com/qdrant/qdrant/pull/6734/commits/f90223018a72dbedd47b1ba19624e10e6298ba5e